server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -185,15 +185,39 @@ private void downloadToFS(String fileName, Path location, String primaryTerm) th
     }
 
     public TranslogTransferMetadata readMetadata() throws IOException {
-        return transferService.listAll(remoteMetadataTransferPath).stream().max(METADATA_FILENAME_COMPARATOR).map(filename -> {
-            try (InputStream inputStream = transferService.downloadBlob(remoteMetadataTransferPath, filename)) {
-                IndexInput indexInput = new ByteArrayIndexInput("metadata file", inputStream.readAllBytes());
-                return metadataStreamWrapper.readStream(indexInput);
-            } catch (IOException e) {
-                logger.error(() -> new ParameterizedMessage("Exception while reading metadata file: {}", filename), e);
-                return null;
-            }
-        }).orElse(null);
+        SetOnce<TranslogTransferMetadata> metadataSetOnce = new SetOnce<>();
+        SetOnce<IOException> exceptionSetOnce = new SetOnce<>();
+        final CountDownLatch latch = new CountDownLatch(1);
+        LatchedActionListener<List<BlobMetadata>> latchedActionListener = new LatchedActionListener<>(
+            ActionListener.wrap(blobMetadataList -> {
+                if (blobMetadataList.isEmpty()) return;
+                String filename = blobMetadataList.get(0).name();
+                try (InputStream inputStream = transferService.downloadBlob(remoteMetadataTransferPath, filename)) {
+                    IndexInput indexInput = new ByteArrayIndexInput("metadata file", inputStream.readAllBytes());
+                    metadataSetOnce.set(metadataStreamWrapper.readStream(indexInput));
+                } catch (IOException e) {
+                    logger.error(() -> new ParameterizedMessage("Exception while reading metadata file: {}", filename), e);
+                    exceptionSetOnce.set(e);
+                }
+            }, e -> {
+                logger.error(() -> new ParameterizedMessage("Exception while listing metadata files "), e);
+                exceptionSetOnce.set((IOException) e);
+            }),
+            latch
+        );
+
+        try {
+            transferService.listAllInSortedOrder(remoteMetadataTransferPath, 1, latchedActionListener);
+            latch.await();
+        } catch (InterruptedException | IOException e) {
+            throw new IOException("Exception while reading/downloading metadafile", e);
```

Comment:
We are consuming most of the exceptions in `exceptionSetOnce` itself. When would we see exception thrown otherwise?

Reply:
we are throwing it later after the read/download finishes .

Reply:
Can we double check what all exceptions can be thrown here?